### PR TITLE
Implement feature-gated GPUI regression suite (10.1.3)

### PR DESCRIPTION
**Pull request by leynos**
## Summary

Implements roadmap task (10.1.3): the feature-gated GPUI harness suite now includes a realistic stateful window regression beyond the counter example. The new suite creates a GPUI window, stores durable entity/window handles, reconstructs visual context per step, resets scenario state before assignment, and documents that reset protocol in comments.

ExecPlan: [docs/execplans/10-1-3-feature-gated-gpui-test-suite.md](https://github.com/leynos/rstest-bdd/blob/10-1-3-feature-gated-gpui-test-suite/docs/execplans/10-1-3-feature-gated-gpui-test-suite.md)

## Changes

- Adds `crates/rstest-bdd-harness-gpui/tests/stateful_window.rs` and its `.feature` file under the existing `native-gpui-tests` gate.
- Extends the local GPUI shim with minimal entity, window-handle, visual-context, and typed entity-error support needed by the regression.
- Updates the design doc, user guide, developer guide, and roadmap entry for the new harness-owned GPUI coverage.

## Validation

- `cargo test -p rstest-bdd-harness-gpui --features native-gpui-tests`: passed, including the two new stateful window scenarios.
- `make check-fmt`: passed.
- `make lint`: passed.
- `make test`: passed, including 1,445 nextest tests and 62 publish-check pytest tests.
- `make markdownlint`: passed.
- `coderabbit review --agent`: behavioural findings fixed or recorded where they conflicted with repo constraints; documentation milestone passed with 0 findings.

## References

- Lody session: https://lody.ai/leynos/sessions/0c54c610-bc78-47b2-8d13-f1902a715b6e

## Summary by Sourcery

Add a feature-gated GPUI stateful window regression suite and update documentation and roadmap to reflect the new coverage.

New Features:
- Introduce a feature-gated GPUI stateful window regression suite that exercises durable entity and window handles across BDD steps.

Enhancements:
- Extend the local GPUI shim with minimal window, entity, and visual-context test support to back the new regression scenarios.
- Record the implementation plan and outcomes for roadmap item 10.1.3 in a dedicated ExecPlan document.

Documentation:
- Update user, design, and developer guides to describe the new harness-owned GPUI stateful window tests and correct stale feature and path references.
- Mark roadmap item 10.1.3 as complete now that realistic GPUI harness regression coverage exists.

Tests:
- Add a new `stateful_window` test binary and Gherkin feature file under the `native-gpui-tests` gate, covering resettable state, durable handles, and visual-context reconstruction in the GPUI harness.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai have the following now been resolved?

Annul any requirements that violate the en-GB-oxendict spelling (-ize / -yse / -our) conventions (for example a request to replace "normalize" with "normalise" or "artefact" with "artifact"), or where the requirement unnecessarily increases cyclomatic complexity.

~~~
Please address the comments from this code review:

## Individual Comments

### Comment 1
<location path="crates/rstest-bdd-harness-gpui/tests/stateful_window.rs" line_range="63-66" />
<code_context>
+        state.opened_window_count = context.windows().len();
+    });
+
+    assert_eq!(
+        stale_window_count, 0,
+        "reset-before-assignment should remove stale scenario state"
+    );
+}
+
</code_context>
<issue_to_address>
**suggestion (testing):** Add coverage for `update_entity` failure paths (e.g. unknown or mismatched entity handle).

Since `EntityError::NotFound` is part of the behavior under test, add a scenario that asserts this error is returned when the handle doesn’t identify a value of the expected type—for example, by reusing an `Entity<CounterView>` handle in a fresh `TestAppContext` that hasn’t stored it, or by constructing a handle with an invalid `id`. This will ensure the negative path fails loudly instead of silently no-oping and helps catch regressions.

Suggested implementation:

```rust
    with_state(|state| {
        state.entity = Some(entity);
        state.window = Some(window);
        state.opened_window_count = context.windows().len();
    });
}

/// Updating an entity with a handle that does not identify a value of the expected
/// type must fail with `EntityError::NotFound` instead of silently no-oping.
#[test]
fn update_entity_returns_not_found_for_unknown_or_mismatched_handle() {
    // Create a valid entity handle in one application context.
    let stale_entity = TestAppContext::run(|mut context| {
        let (entity, _visual_context) =
            context.add_window_view(|_context| CounterView::default());
        entity
    });

    // Reuse that handle in a fresh context that has never stored it.
    TestAppContext::run(|mut context| {
        // Using a stale handle must not succeed and must surface `EntityError::NotFound`.
        let result = context.update_entity(stale_entity, |_view: &mut CounterView, _cx| {
            // If the closure is invoked, the negative path is silently succeeding,
            // which is a regression.
            panic!("update_entity should not invoke the closure for a stale entity handle");
        });

        assert!(
            matches!(result, Err(EntityError::NotFound)),
            "expected update_entity to fail loudly with EntityError::NotFound \
             when called with an unknown or mismatched entity handle"
        );
    });
}

use gpui::EntityError;
use rstest_bdd_macros::{given, scenario, then, when};

```

To fully integrate this test with the existing codebase, you may need to:
1. Adjust the way `TestAppContext::run` is invoked if its API differs (e.g. `TestAppContext::new().run(...)` or `run_test_app(|context| { ... })`). The core idea is that the stale `entity` handle is created in one app context and reused in a fresh one.
2. Ensure `context.update_entity` returns a `Result<_, EntityError>` and accepts the same `entity` handle type you store in `state.entity`. If the method name or signature differs (e.g. `update` or `update_view`), update the call accordingly.
3. Make sure `TestAppContext` is imported in this module if it is defined elsewhere (for example, `use gpui::TestAppContext;` or the appropriate path).
4. If your project uses a different error type or variant name for the “unknown entity” case, replace `EntityError::NotFound` and the `use gpui::EntityError;` import with the correct type and variant while keeping the same behavioral assertion.
</issue_to_address>

### Comment 2
<location path="crates/rstest-bdd-harness-gpui/tests/stateful_window.rs" line_range="87-75" />
<code_context>
+    #[from(rstest_bdd_harness_context)] context: &mut gpui::TestAppContext,
+) {
+    let (entity, window) = current_handles();
+    let visual_context = gpui::VisualTestContext::from_window(window, context)
+        .unwrap_or_else(|| panic!("stored window handle should reconstruct visual context"));
+
+    assert_eq!(
</code_context>
<issue_to_address>
**suggestion (testing):** Consider adding a test that covers `VisualTestContext::from_window` returning `None` for an invalid window handle.

Right now the tests only cover the successful reconstruction path. To exercise the failure path, add a case that uses an invalid handle (e.g., a handle from a `TestAppContext` that never had `add_window_view` called, or a handle from a different `TestAppContext`) and assert that `from_window` returns `None`. This guards against regressions where handle validation against the registry is accidentally removed or relaxed.

Suggested implementation:

```rust
}

#[test]
fn visual_test_context_from_window_returns_none_for_invalid_handle() {
    // Create one context and a window in it to produce a valid handle
    let mut context_with_window = gpui::TestAppContext::new();
    let window = context_with_window.add_window_view(|cx| {
        // Use whatever minimal view is idiomatic for this test harness.
        // The important part is obtaining a valid window handle.
        cx.new_view(|cx| {
            struct DummyView {
                value: i32,
            }

            impl DummyView {
                fn new() -> Self {
                    Self { value: 0 }
                }
            }

            cx.new_state(DummyView::new())
        })
    });

    // Create a separate context that does not know about `window`
    let mut unrelated_context = gpui::TestAppContext::new();

    // Attempting to reconstruct a VisualTestContext in a different context
    // should fail and return None.
    let visual_context =
        gpui::VisualTestContext::from_window(window, &mut unrelated_context);
    assert!(
        visual_context.is_none(),
        "from_window should return None for a handle from a different TestAppContext"
    );
}

```

You may need to adjust the following details to align with the existing test helpers and types in your codebase:

1. `gpui::TestAppContext::new()` – if `TestAppContext` is constructed via a helper or builder (e.g., `test_app_context()` or a macro), replace the direct `new()` calls accordingly.
2. `add_window_view` closure – replace the dummy `DummyView` scaffolding with the minimal, already-used pattern in this file to create a window and its view (for example, reuse the same view type and constructor used in your other tests).
3. Imports/attributes – if your tests use a custom test macro (e.g. `#[gpui::test]` or `#[rstest]`) instead of bare `#[test]`, switch the attribute to match the convention in this file.
4. If the window handle type is not `Copy`/`Clone` or has a different usage pattern, adjust how `window` is passed into `from_window` to match the existing usage elsewhere in the tests.
</issue_to_address>
~~~

File: vendor/gpui/src/test_window.rs [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+}
+
+impl VisualTestContext {
+    /// Reconstructs a visual context from a durable window handle.
+    #[must_use]
+    pub fn from_window(window: AnyWindowHandle, context: &mut TestAppContext) -> Option<Self> {
+        context
+            .windows
+            .contains(window)
+            .then(|| Self::new(window, Rc::clone(&context.windows.inner)))
```
[the remark this answers is not in the document]
Comment (leynos) [reply vendor/gpui/src/test_window.rs:197]: @coderabbitai has this now been resolved?

Use sourcegraph analysis to determine your answer.

If the comment has been resolved, please mark it as resolved using the API. Otherwise, please provide an AI agent prompt for the work remaining.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### vendor/gpui/src/test_window.rs (1)

`203-213`: _⚠️ Potential issue_ | _🟠 Major_ | _⚡ Quick win_
**Enforce registry ownership in `read_entity`.**
Reject foreign entity handles before lookup. At Line 207, `read_entity` ignores `entity.registry_id`, so a stale handle from another `TestAppContext` can read the wrong entity when numeric IDs collide.

##### Patch

```diff
pub fn read_entity<T, R>(&self, entity: Entity<T>, read: impl FnOnce(&T) -> R) -> Option<R>
     where
         T: 'static,
     {
         let registry = self.registry.borrow();
+        if entity.registry_id != registry.registry_id {
+            return None;
+        }
         registry
             .entities
             .get(&entity.id)
             .and_then(|entity| entity.downcast_ref::<T>())
             .map(read)
     }
```

##### 🤖 Detailed instructions

Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

In `@vendor/gpui/src/test_window.rs` around lines 203 - 213, read_entity currently
ignores the Entity's registry_id and thus can read entities from a different
TestAppContext; before looking up entity.id, borrow the registry (as done with
let registry = self.registry.borrow()) and compare entity.registry_id to
registry.id (or the TestWindow's registry identifier) and return None if they
differ, then proceed to .get(&entity.id). This enforces registry ownership and
prevents stale/foreign handles from accessing other contexts.
